{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Complement: AATGCT\n",
      "mRNA: AAUGCU\n",
      "Protein: Asn-Ala\n"
     ]
    }
   ],
   "source": [
    "#Assignment 1, Louis Ruby Elsalim 2602185431 \n",
    "\n",
    "# DNA to mRNA codon mapping\n",
    "codon_table = {\n",
    "    'AUG': 'Met', 'UUU': 'Phe', 'UUC': 'Phe', 'UUA': 'Leu', 'UUG': 'Leu', \n",
    "    'CUU': 'Leu', 'CUC': 'Leu', 'CUA': 'Leu', 'CUG': 'Leu', 'AUU': 'Ile', \n",
    "    'AUC': 'Ile', 'AUA': 'Ile', 'GUU': 'Val', 'GUC': 'Val', 'GUA': 'Val', \n",
    "    'GUG': 'Val', 'UCU': 'Ser', 'UCC': 'Ser', 'UCA': 'Ser', 'UCG': 'Ser', \n",
    "    'CCU': 'Pro', 'CCC': 'Pro', 'CCA': 'Pro', 'CCG': 'Pro', 'ACU': 'Thr', \n",
    "    'ACC': 'Thr', 'ACA': 'Thr', 'ACG': 'Thr', 'GCU': 'Ala', 'GCC': 'Ala', \n",
    "    'GCA': 'Ala', 'GCG': 'Ala', 'UAU': 'Tyr', 'UAC': 'Tyr', 'UGU': 'Cys', \n",
    "    'UGC': 'Cys', 'UGG': 'Trp', 'CAA': 'Gln', 'CAG': 'Gln', 'AAU': 'Asn', \n",
    "    'AAC': 'Asn', 'GAU': 'Asp', 'GAC': 'Asp', 'GAA': 'Glu', 'GAG': 'Glu', \n",
    "    'CAU': 'His', 'CAC': 'His', 'AAA': 'Lys', 'AAG': 'Lys', 'CGU': 'Arg', \n",
    "    'CGC': 'Arg', 'CGA': 'Arg', 'CGG': 'Arg', 'AGA': 'Arg', 'AGG': 'Arg', \n",
    "    'GGU': 'Gly', 'GGC': 'Gly', 'GGA': 'Gly', 'GGG': 'Gly', 'UAA': 'Stop', \n",
    "    'UAG': 'Stop', 'UGA': 'Stop'\n",
    "}\n",
    "\n",
    "# Function to translate DNA to mRNA and then to protein\n",
    "def translate_dna_to_protein(dna_sequence):\n",
    "    # Make sure that the length of the DNA sequence is a multiple of 3\n",
    "    if len(dna_sequence) % 3 != 0:\n",
    "        raise ValueError(\"The input DNA sequence must be a multiple of 3 in length.\")\n",
    "    \n",
    "    # DNA complement (A->T, T->A, C->G, G->C)\n",
    "    complement = dna_sequence.replace('A', 't').replace('T', 'a').replace('C', 'g').replace('G', 'c').upper()\n",
    "    \n",
    "    # Transcription: DNA to mRNA (T -> U)\n",
    "    mrna = complement.replace('T', 'U')\n",
    "    \n",
    "    # Translate mRNA to Protein\n",
    "    protein_sequence = []\n",
    "    for i in range(0, len(mrna), 3):\n",
    "        codon = mrna[i:i+3]\n",
    "        amino_acid = codon_table.get(codon, 'X')  # 'X' is for unknown codon\n",
    "        if amino_acid == 'Stop':\n",
    "            break\n",
    "        protein_sequence.append(amino_acid)\n",
    "    \n",
    "    # Join protein sequence with \"-\" if there are multiple amino acids\n",
    "    protein_sequence_formatted = '-'.join(protein_sequence)\n",
    "    \n",
    "    return complement, mrna, protein_sequence_formatted\n",
    "\n",
    "# Test with input DNA sequence\n",
    "dna_sequence = \"TTACGA\" # Input DNA sequence here\n",
    "complement, mrna, protein = translate_dna_to_protein(dna_sequence)\n",
    "print(\"Complement:\", complement)\n",
    "print(\"mRNA:\", mrna)\n",
    "print(\"Protein:\", protein)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amino acid sequence: WYW\n",
      "Possible mRNA sequences and their codon frequencies:\n",
      "\n",
      "mRNA sequence: UGGUAUUGG\n",
      "Codon frequencies:\n",
      "UGG = 2\n",
      "UAU = 1\n",
      "\n",
      "mRNA sequence: UGGUACUGG\n",
      "Codon frequencies:\n",
      "UGG = 2\n",
      "UAC = 1\n"
     ]
    }
   ],
   "source": [
    "#Assignment 2\n",
    "\n",
    "import itertools\n",
    "\n",
    "# Codon to Amino Acid mapping\n",
    "codon_table = {\n",
    "    'UUU': 'F', 'UUC': 'F', 'UUA': 'L', 'UUG': 'L',\n",
    "    'UCU': 'S', 'UCC': 'S', 'UCA': 'S', 'UCG': 'S',\n",
    "    'UAU': 'Y', 'UAC': 'Y', 'UAA': 'STOP', 'UAG': 'STOP',\n",
    "    'UGU': 'C', 'UGC': 'C', 'UGA': 'STOP', 'UGG': 'W',\n",
    "    'CUU': 'L', 'CUC': 'L', 'CUA': 'L', 'CUG': 'L',\n",
    "    'CCU': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',\n",
    "    'CAU': 'H', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "    'CGU': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R',\n",
    "    'AUU': 'I', 'AUC': 'I', 'AUA': 'I', 'AUG': 'M',\n",
    "    'ACU': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',\n",
    "    'AAU': 'N', 'AAC': 'N', 'AAA': 'K', 'AAG': 'K',\n",
    "    'AGU': 'S', 'AGC': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "    'GUU': 'V', 'GUC': 'V', 'GUA': 'V', 'GUG': 'V',\n",
    "    'GCU': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',\n",
    "    'GAU': 'D', 'GAC': 'D', 'GAA': 'E', 'GAG': 'E',\n",
    "    'GGU': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G'\n",
    "}\n",
    "\n",
    "# Reverse the codon table to map amino acids to codons\n",
    "amino_acid_to_codons = {}\n",
    "for codon, amino_acid in codon_table.items():\n",
    "    if amino_acid not in amino_acid_to_codons:\n",
    "        amino_acid_to_codons[amino_acid] = []\n",
    "    amino_acid_to_codons[amino_acid].append(codon)\n",
    "\n",
    "# Function to generate all possible mRNA sequences from an amino acid sequence\n",
    "def codon_combinations(amino_acid_sequence):\n",
    "    # List all possible codons for each amino acid\n",
    "    possible_codons = [amino_acid_to_codons[aa] for aa in amino_acid_sequence if aa in amino_acid_to_codons]\n",
    "    \n",
    "    # Generate all possible combinations of codons using Cartesian product\n",
    "    possible_mRNA_sequences = [''.join(codon_set) for codon_set in itertools.product(*possible_codons)]\n",
    "    \n",
    "    return possible_mRNA_sequences\n",
    "\n",
    "# Function to calculate codon frequency in a given mRNA sequence\n",
    "def codon_frequency(mRNA_sequence):\n",
    "    codon_count = {}\n",
    "    for i in range(0, len(mRNA_sequence) - 2, 3):\n",
    "        codon = mRNA_sequence[i:i+3]\n",
    "        if codon in codon_count:\n",
    "            codon_count[codon] += 1\n",
    "        else:\n",
    "            codon_count[codon] = 1\n",
    "    return codon_count\n",
    "\n",
    "# Function to process amino acid sequence and print all results\n",
    "def codon_frequency_from_amino_acids(amino_acid_sequence):\n",
    "    # Generate all possible mRNA sequences\n",
    "    possible_mRNA_sequences = codon_combinations(amino_acid_sequence)\n",
    "    \n",
    "    print(f\"Amino acid sequence: {amino_acid_sequence}\")\n",
    "    print(\"Possible mRNA sequences and their codon frequencies:\")\n",
    "    \n",
    "    for mRNA_sequence in possible_mRNA_sequences:\n",
    "        print(f\"\\nmRNA sequence: {mRNA_sequence}\")\n",
    "        codon_count = codon_frequency(mRNA_sequence)\n",
    "        print(\"Codon frequencies:\")\n",
    "        for codon, count in codon_count.items():\n",
    "            print(f\"{codon} = {count}\")\n",
    "\n",
    "# Example usage\n",
    "amino_acid_sequence = \"WYW\"  # Example amino acids sequence\n",
    "codon_frequency_from_amino_acids(amino_acid_sequence)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
